File: Services/influx.py
from os import path
import json
from datetime import datetime, timedelta
from influxdb_client import InfluxDBClient, Point
from influxdb_client.client.write_api import SYNCHRONOUS

class Influx:
    def __init__(self):
        f = open(path.abspath(path.join(path.dirname(__file__), '../Influx/influx-config.json')), 'r')
        self.influx_config = json.load(f)
        f.close()
        f = open(path.abspath(path.join(path.dirname(__file__), '../Weather/weather-service-config.json')), 'r')
        self.weather_config = json.load(f)
        f.close()
        f = open(path.abspath(path.join(path.dirname(__file__), '../Forecasting/forecasting-service-config.json')), 'r')
        self.forecasting_config = json.load(f)
        f.close()
        self.client = InfluxDBClient(url=self.influx_config['url'], token=self.influx_config['token'], org=self.influx_config['organization'])
        self.write_api = self.client.write_api(write_options=SYNCHRONOUS)
        self.query_api = self.client.query_api()
        self.delete_api = self.client.delete_api()

    def write_weather(self, point_dict):
        print('INFLUX -> [WAIT] write weather')
        try:
            point_dict['measurement'] = self.influx_config['measurement']['weather']
            point_dict['tags'] = self.weather_config
            self.write_api.write(bucket=self.influx_config['bucket']['weather'], org=self.influx_config['organization'], record=point_dict)
            print('INFLUX -> [OK  ] write weather')
        except:
            print('INFLUX -> [ERR ] write weather')

    def write_forecasting_arima(self, point_dict):
        print('INFLUX -> [WAIT] write forecasting arima')
        try:
            point_dict['measurement'] = self.influx_config['measurement']['forecasting']['arima']
            point_dict['tags'] = self.forecasting_config
            self.write_api.write(bucket=self.influx_config['bucket']['sensor'], org=self.influx_config['organization'], record=point_dict)
            print('INFLUX -> [OK  ] write forecasting arima')
        except:
            print('INFLUX -> [ERR ] write forecasting arima')

    def write_forecasting_prophet(self, point_dict):
        print('INFLUX -> [WAIT] write forecasting prophet')
        try:
            point_dict['measurement'] = self.influx_config['measurement']['forecasting']['prophet']
            point_dict['tags'] = self.forecasting_config
            self.write_api.write(bucket=self.influx_config['bucket']['sensor'], org=self.influx_config['organization'], record=point_dict)
            print('INFLUX -> [OK  ] write forecasting prophet')
        except:
            print('INFLUX -> [ERR ] write forecasting prophet')

    def delete_forecasting_arima_all(self):
        print('INFLUX -> [WAIT] delete forecasting arima')
        try:
            self.delete_api.delete('1970-01-01T00:00:00Z', (datetime.now() + timedelta(days=365)).strftime('%Y-%m-%dT%H:%M:%SZ'), '_measurement="{:s}"'.format(self.influx_config['measurement']['forecasting']['arima']), bucket=self.influx_config['bucket']['sensor'], org=self.influx_config['organization'])
            print('INFLUX -> [OK  ] delete forecasting arima')
        except Exception as e:
            print('INFLUX -> [ERR ] delete forecasting arima')
            print(e)

    def delete_forecasting_prophet_all(self):
        print('INFLUX -> [WAIT] delete forecasting prophet')
        try:
            self.delete_api.delete('1970-01-01T00:00:00Z', (datetime.now() + timedelta(days=365)).strftime('%Y-%m-%dT%H:%M:%SZ'), '_measurement="{:s}"'.format(self.influx_config['measurement']['forecasting']['prophet']), bucket=self.influx_config['bucket']['sensor'], org=self.influx_config['organization'])
            print('INFLUX -> [OK  ] delete forecasting prophet')
        except Exception as e:
            print('INFLUX -> [ERR ] delete forecasting prophet')
            print(e)
        

File: Services/influx.js
const fs = require('fs');
const path = require('path');
const {InfluxDB, Point} = require('@influxdata/influxdb-client');

const config = JSON.parse(fs.readFileSync(path.join(__dirname, '../Influx/influx-config.json'), 'utf-8'));

module.exports = {
  _config: config,
  _influxdb: new InfluxDB({
    url: config.url,
    token: config.token
  }),
  writeSensor: function(dataStr) {
    console.log('INFLUX -> [WAIT] write sensor');
    try {
      let dataArr = dataStr.split('|');
      let point = new Point('sensor')
        .tag('id', dataArr[0])
        .tag('latitude', dataArr[1])
        .tag('longitude', dataArr[2])
        .intField('rssi', dataArr[3])
        .intField('packets', dataArr[4])
        .intField('delay', dataArr[5])
        .floatField('temperature', dataArr[6])
        .floatField('humidity', dataArr[7])
        .intField('gas', dataArr[8])
        .intField('aqi', dataArr[9]);
      const writeApi = this._influxdb.getWriteApi(config.organization, this._config.bucket.sensor);
      writeApi.writePoint(point);
      writeApi.close()
        .then((result) => {
          console.log('INFLUX -> [OK  ] write sensor');
        })
        .catch((error) => {
          console.log('INFLUX -> [ERR ] write sensor');
        });
    } catch(error) {
      console.log('INFLUX -> [ERR ] write sensor');
    }
  }
};
